libs/analysis/bbhnet/analysis/vt.py

```diff
@@ -0,0 +1,163 @@
+    dec_min, dec_max = dec_min * u.rad, dec_max * u.rad
+    theta_max, theta_min = (
+        np.pi / 2 - dec_min.value,
+        np.pi / 2 - dec_max.value,
+    )
+    omega = -2 * math.pi * (np.cos(theta_max) - np.cos(theta_min))
+
+    # calculate the volume of the universe
+    # over which injections have been made
+    integrand = (
+        lambda z: 1.0
+        / (1 + z)
+        * (cosmology.differential_comoving_volume(z)).value
+    )
+    volume, _ = quad(integrand, zmin, zmax) * u.Mpc**3 * omega
+    return volume
+
+
+@dataclass
+class VolumeTimeIntegral:
+    """
+    Class for calculating VT metrics using importance sampling.
+
+    Args:
+        source:
+            Bilby PriorDict of the source distribution
+            used to create injections
+        recovered_parameters:
+            Dictionary of recovered parameters
+        n_injections:
+            Number of total injections
+        livetime:
+            Livetime in seconds over which injections were performed
+        cosmology:
+            Astropy Cosmology object used for volume calculation
+    """
+
+    source: "bilby.core.prior.PriorDict"
+    recovered_parameters: Dict[str, np.ndarray]
+    n_injections: int
+    livetime: float
+    cosmology: "Cosmology" = cosmo.Planck15
+
+    def __post_init__(self):
+        # convert recovered parameters to a list of dictionaries
+        self.recovered_parameters = [
+            dict(zip(self.recovered_parameters, col))
+            for col in zip(*self.recovered_parameters.values())
+        ]
+
+        dl_prior = self.source["luminosity_distance"]
+        dl_min, dl_max = [dl_prior.minimum, dl_prior.maximum]
+
+        # if the source distribution has a dec prior,
+        # use it to calculate the area on the sky
+        # over which injections have been made
+        # otherwise, calculate_astrophysical_volume assumes the full sky
+        if "dec" in self.source:
+            dec_prior = self.source["dec"]
+            dec_min, dec_max = dec_prior.minimum, dec_prior.maximum
+        else:
+            dec_min = dec_max = None
+
+        # calculate the astrophysical volume over
+        # which injections have been made.
+        self.volume = calculate_astrophysical_volume(
+            dl_min=dl_min,
+            dl_max=dl_max,
+            dec_min=dec_min,
+            dec_max=dec_max,
+            cosmology=self.cosmology,
+        )
+
+    def weights(self, target: Optional["bilby.core.prior.PriorDict"] = None):
+        """
+        Calculate the weights for the samples.
+        """
+
+        # if no target distribution is passed,
+        # use the source distribution
+        if target is None:
+            target = self.source
+
+        weights = []
+        for sample in self.recovered_parameters:
+            # calculate the weight for each sample
+            # using the source and target distributions
+            weight = target.prob(sample) / self.source.prob(sample)
```

Comment:
I think this might be fine. My concern was the dimensions not matching. But just did a quick check.

```
prior_a = bilby.prior.Uniform(1, 101)
prior_b = bilby.prior.Uniform(1, 101)
prior_c = bilby.prior.Uniform(0, 0.5)

source = bilby.prior.PriorDict(dict(a=prior_a, b=prior_b))
target = bilby.prior.PriorDict(dict(a=prior_a, b=prior_b, c=prior_c))

>>> source.prob(dict(a=5, b=6))
0.0001
>>> target.prob(dict(a=5, b=6))
0.0001
```
It looks like if the target prior has more number of dimensions that `recovered_parameters` it will marginalize over those.

However, to avoid confusion/assumptions, it might be better to pass only the `weighting_parameter_keys`, which is assumed to be present in both
```
source_subset = PriorDict({k: self.source[k] for k in weighting_parameter_keys})
target_subset = PriorDict({k: target[k] for k in weighting_parameter_keys})

weight = target_subset(sample) / source_subset(sample)
```
```

Reply:
Awesome thanks for looking into this. Makes sense